### PR TITLE
Support kubevirt test run on sle-micro

### PR DESCRIPTION
**Pull request by nanzhg**
1. Create a separate yaml file "slem-kubevirt-tests.yaml" for kubevirt test on sle-micro
2. Introduce a new variable 'KUBEVIRT_TEST' for running full or core tests
3. Split into two functions for installing kubevirt packages and deploying kubevirt manifests
4. Add kubevirt full & core tests configuration files
    * data/virt_autotest/kubevirt_tests/full-tests.conf
    * data/virt_autotest/kubevirt_tests/core-tests.conf
5. Update test mode prepare_transactional_server.pm to leverage in kubevirt test on sle-micro
    * Fix package declaration issue
    * Extra the common part as a function "install_additional_pkgs()" for leverage in kubevirt server setup

- Related ticket: https://progress.opensuse.org/issues/113228
- Verification run: 
    * For sle-micro - http://10.67.129.66/tests/1672, http://10.67.129.66/tests/1729
    * For sle - http://10.67.129.66/tests/1735

### CODE REVIEW

**Comment by nanzhg** [comment]
@alice-suse @xguo @waynechen55 @Julie-CAO @tbaev @tonyyuan1 @RoyCai7 please help to review this PR. Thanks

File: data/virt_autotest/kubevirt_tests/core-tests.conf [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# SUSE's openQA tests
+#
+# Copyright 2022 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Kubevirt core test configuration
+# Maintainer: Nan Zhang <nan.zhang@suse.com>
```
Comment (alice-suse) [inline data/virt_autotest/kubevirt_tests/core-tests.conf:7]: Would you please also add our team ML to maintainer, qe-virt@suse.de?

File: data/virt_autotest/kubevirt_tests/full-tests.conf [hunk 2]
```diff
@@ -3,7 +3,7 @@
 # Copyright 2022 SUSE LLC
 # SPDX-License-Identifier: FSFAP
 
-# Summary: The list of kubevirt test suites and parameters configuration
+# Summary: Kubevirt full test configuration
 # Maintainer: Nan Zhang <nan.zhang@suse.com>
```
Comment (alice-suse) [inline data/virt_autotest/kubevirt_tests/full-tests.conf:7]: Same here, maintainer info

File: tests/virt_autotest/kubevirt_tests_agent.pm [hunk 3]
```diff
@@ -12,13 +12,16 @@ use strict;
...
 use utils;
+use version_utils qw(is_sle_micro);
+use Utils::Systemd;
 use Utils::Backends 'use_ssh_serial_console';
 
 sub run {
     my ($self) = shift;
 
-    if (get_required_var('WITH_SLE_INSTALL')) {
+    if (get_var('WITH_SLE_INSTALL') || get_var('WITH_SLEM_INSTALL')) {
```
Comment (alice-suse) [inline tests/virt_autotest/kubevirt_tests_agent.pm:24]: Are WITH_SLE_INSTALL and WITH_SLEM_INSTALL indicators for host os? If yes, maybe is_sle/is_sle_micro are more common way?

File: tests/virt_autotest/kubevirt_tests_agent.pm [hunk 4]
```diff
@@ -43,15 +46,22 @@ sub rke2_agent_setup {
...
+        disable_and_stop_service('firewalld.service');
+    }
     $self->setup_passwordless_ssh_login($server_ip);
 
-    # Ensure SUSE certificates are installed on the node
+    # rebootmgr has to be turned off as prerequisity for this to work
+    script_run "rebootmgrctl set-strategy off" if (is_sle_micro);
+    # Check if the package 'ca-certificates-suse' are installed on the node
     ensure_ca_certificates_suse_installed();
+    transactional::process_reboot(trigger => 1) if (is_sle_micro);
```
Comment (alice-suse) [inline tests/virt_autotest/kubevirt_tests_agent.pm:59]: Maybe is_transactional is better, to avoid change for alp in future.

File: tests/virt_autotest/kubevirt_tests_server.pm [hunk 5]
```diff
@@ -84,15 +99,22 @@ sub rke2_server_setup {
...
     # Check if the package 'ca-certificates-suse' are installed on the node
     ensure_ca_certificates_suse_installed();
+    transactional::process_reboot(trigger => 1) if (is_sle_micro);
 
-    # Install downstream kubevirt packages
-    zypper_call('in kubernetes1.18-client') if (script_run('rpmquery kubernetes1.18-client'));
+    unless (is_sle_micro) {
+        # Install downstream kubevirt packages
+        zypper_call('in kubernetes1.18-client') if (script_run('rpmquery kubernetes1.18-client'));
+    }
```
Comment (alice-suse) [inline tests/virt_autotest/kubevirt_tests_server.pm:117]: The above section ,line 102 until here, seems to be the same with agent.pm. Have you considered wrapping it into a function and share by agent and server?

File: tests/virt_autotest/kubevirt_tests_server.pm [hunk 6]
```diff
@@ -181,13 +208,26 @@ sub install_kubevirt_pkgs_n_deploy_manifests {
     if ($virt_manifests_repo) {
         zypper_call("in -f -r Virt-Manifests-Repo $virt_manifests");
     } elsif (script_run("rpmquery $virt_manifests")) {
-        zypper_call("in -f -r SLE-Module-Containers${os_version}-Updates $virt_manifests");
+        if (is_sle_micro) {
+            zypper_call("in -f $virt_manifests");
+        } else {
+            zypper_call("in -f -r SLE-Module-Containers${os_version}-Updates $virt_manifests");
```
Comment (alice-suse) [inline tests/virt_autotest/kubevirt_tests_server.pm:214]: What if SLE-Module-Containers${os_version}-Updates does not have update for manifest. eg sle15sp5? Zypper will automatically install the highest version, right? Why is here needed to specify repo?

File: tests/virt_autotest/kubevirt_tests_server.pm [hunk 7]
```diff
@@ -241,12 +291,17 @@ sub run_virt_tests {
     assert_script_run("mkdir -p $result_dir");
 
     # Run virt-tests command for each go test files
-    my $test_conf = 'kubevirt-tests.conf';
+    if (check_var('KUBEVIRT_TEST_SUITE', 'full')) {
+        $test_conf = 'full-tests.conf';
+        @kubevirt_tests = @full_tests;
+    } elsif (check_var('KUBEVIRT_TEST_SUITE', 'core')) {
```
Comment (alice-suse) [inline tests/virt_autotest/kubevirt_tests_server.pm:297]: A separate core list is a workable way. I do not intend to suggest changing it. Just that I do not figure out why you do not directly use the way to customize test list by giving test_ids in ginkgo filter list, which is supported in existing automation already, right? Any reasons?

File: tests/virt_autotest/transactional/prepare_transactional_server.pm [hunk 8]
```diff
@@ -9,7 +9,7 @@
 # only or which need to to performed by leveraging transactional-update command.
 #
 # Maintainer: Wayne Chen <wchen@suse.com> qe-virt@suse.de
-package prepare_transactional_server;
+package transactional::prepare_transactional_server;
```
Comment (alice-suse) [inline tests/virt_autotest/transactional/prepare_transactional_server.pm:12]: Good catch!

**Review by alice-suse — COMMENTED** [review]
Overall it looks good to me. Just some minor comments given. Nice job!

File: tests/virt_autotest/kubevirt_tests_agent.pm (hunk 4 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_agent.pm:59]: Good idea! Already updated.

File: tests/virt_autotest/kubevirt_tests_agent.pm (hunk 3 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_agent.pm:24]: This parameter is just used for controlling if the test run with fresh OS installation or running on an installed system.

File: data/virt_autotest/kubevirt_tests/core-tests.conf (hunk 1 above)
Comment (nanzhg) [reply data/virt_autotest/kubevirt_tests/core-tests.conf:7]: Yes, updated.

File: data/virt_autotest/kubevirt_tests/full-tests.conf (hunk 2 above)
Comment (nanzhg) [reply data/virt_autotest/kubevirt_tests/full-tests.conf:7]: Updated.

File: tests/virt_autotest/kubevirt_tests_server.pm (hunk 5 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm:117]: It's actually specified for kubevirt test, not very general for most cases, so I would not encapsulate into a function. In addition, frequent function call will increase the system overhead.

Comment (alice-suse) [reply tests/virt_autotest/kubevirt_tests_server.pm:117]: okay, fine with it if it is your decision

File: schedule/virt_autotest/slem-kubevirt-tests.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    - '{{barrier_setup}}'
+    - '{{bootup_and_install}}'
+    - '{{kubevirt_tests}}'
+conditional_schedule:
+    barrier_setup:
+        TEST_SUITE_NAME:
+            slem-kubevirt-tests-server:
+                - virt_autotest/kubevirt_barriers
+    bootup_and_install:
+        WITH_SLEM_INSTALL:
```
Comment (Julie-CAO) [inline schedule/virt_autotest/slem-kubevirt-tests.yaml:19]: you have different yaml files to schedule SLE and SLEM tests, I guess it is because of different host installation steps. It is acceptable but you will have to maintain two files in the future.

File: schedule/virt_autotest/slem-kubevirt-tests.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            1:
+                - installation/ipxe_install
+            0:
+                - boot/boot_from_pxe
+    kubevirt_tests:
+        TEST_SUITE_NAME:
+            slem-kubevirt-tests-server:
+                - virt_autotest/kubevirt_tests_server
+            slem-kubevirt-tests-agent:
+                - virt_autotest/kubevirt_tests_agent
```
Comment (waynechen55) [inline schedule/virt_autotest/slem-kubevirt-tests.yaml:47]: I think you can try ```yq``` command to validate your yaml file format and syntax better.

File: tests/virt_autotest/kubevirt_tests_agent.pm (hunk 3 above)
Comment (waynechen55) [reply tests/virt_autotest/kubevirt_tests_agent.pm:24]: Then I think you should use WITH_HOST_INSTALL.

File: tests/virt_autotest/kubevirt_tests_agent.pm [hunk 11]
```diff
@@ -43,15 +46,22 @@ sub rke2_agent_setup {
...
+    # Check if the package 'ca-certificates-suse' are installed on the node
     ensure_ca_certificates_suse_installed();
+    transactional::process_reboot(trigger => 1) if (is_transactional);
 
-    # Install downstream kubevirt packages
-    zypper_call('in -n kubernetes1.18-client') if (script_run('rpmquery kubernetes1.18-client'));
+    unless (is_transactional) {
+        # Install downstream kubevirt packages
+        zypper_call('in -n kubernetes1.18-client') if (script_run('rpmquery kubernetes1.18-client'));
+    }
```
Comment (waynechen55) [inline tests/virt_autotest/kubevirt_tests_agent.pm:65]: I think you can group your codes into "transactional" and "non-transactional" blocks.

File: schedule/virt_autotest/slem-kubevirt-tests.yaml (hunk 9 above)
Comment (nanzhg) [reply schedule/virt_autotest/slem-kubevirt-tests.yaml:19]: Updated to use one yaml file. Thank you for suggestion.

File: tests/virt_autotest/kubevirt_tests_agent.pm (hunk 3 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_agent.pm:24]: Updated. Thank you for suggestion.

File: schedule/virt_autotest/slem-kubevirt-tests.yaml (hunk 10 above)
Comment (nanzhg) [reply schedule/virt_autotest/slem-kubevirt-tests.yaml:47]: Thanks for advise! I got green for my yaml file by using this command.

File: tests/virt_autotest/kubevirt_tests_agent.pm (hunk 11 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_agent.pm:65]: I would like to remove this part. The package is not required currently, as the kubectl command is provided by rke2 installation.

File: schedule/virt_autotest/slem-kubevirt-tests.yaml [hunk 12]
```diff
@@ -0,0 +1,47 @@
+name:         slem-kubevirt-tests
+description:  >
+    Maintainer: Nan Zhang <nan.zhang@suse.com> qe-virt@suse.de
+    Run kubevirt tests on SLE Micro
+vars:
+    VIDEOMODE: text
+    SYSTEM_ROLE: kvm
+    OPT_KERNEL_PARAMS: hugepagesz=2M hugepages=64 max_loop=256
```
Comment (waynechen55) [inline schedule/virt_autotest/slem-kubevirt-tests.yaml:8]: Have you verified these vars put here ? Do they really take effect ?

File: tests/virt_autotest/kubevirt_tests_agent.pm (hunk 11 above)
Comment (waynechen55) [reply tests/virt_autotest/kubevirt_tests_agent.pm:65]: I think you can organize them better if possible. Now it seems small code blocks litter everywhere.

File: tests/virt_autotest/kubevirt_tests_server.pm (hunk 6 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm:214]: The default repo is from the build installer, and the variable VIRT_MANIFESTS_REPO is for testing the kubevirt packages from devel repo or others. We need to test kubevirt packages along with the build.

File: schedule/virt_autotest/slem-kubevirt-tests.yaml (hunk 12 above)
Comment (nanzhg) [reply schedule/virt_autotest/slem-kubevirt-tests.yaml:8]: Yes, they are verified and they were moved out from yaml file in the latest comment.

File: tests/virt_autotest/kubevirt_tests_server.pm (hunk 7 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm:300]: Well, as the test ids were picked up from different test classes, I'd like to generate a test report for those test ids under matched test classes, so that we can know what the test comes from.

File: tests/virt_autotest/kubevirt_tests_agent.pm (hunk 11 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_agent.pm:65]: The snippet of code is useless and removed, please review again. Thanks

File: tests/virt_autotest/kubevirt_tests_server.pm (hunk 6 above)
Comment (alice-suse) [reply tests/virt_autotest/kubevirt_tests_server.pm:209]: I know the story that during 15sp4, it is needed to install the pkg from SLE-Module-Containers${os_version}-Updates, but that's an exception, and it won't always be the case.

File: tests/virt_autotest/kubevirt_tests_server.pm (hunk 7 above)
Comment (alice-suse) [reply tests/virt_autotest/kubevirt_tests_server.pm:295]: I see. Make sense :)

File: tests/virt_autotest/transactional/prepare_transactional_server.pm [hunk 13]
```diff
@@ -69,7 +75,7 @@ sub prepare_packages {
         # should not prevent user from installing legitimate tools and utilities, it
         # is expected that use of additional packages should be limited to the minimum
         # and their impact should be analyzed beforehand.
-        my @repos_to_install = split(/,/, get_var("INSTALL_OTHER_REPOS"));
+        my @repos_to_install = split(/,/, $additional_repos);
```
Comment (waynechen55) [inline tests/virt_autotest/transactional/prepare_transactional_server.pm:78]: Why this change ?

File: tests/virt_autotest/transactional/prepare_transactional_server.pm [hunk 14]
```diff
@@ -82,7 +88,7 @@ sub prepare_packages {
         save_screenshot;
 
         my $cmd = "--non-interactive install --no-allow-downgrade --no-allow-name-change --no-allow-vendor-change";
-        $cmd = $cmd . " $_" foreach (split(/,/, get_required_var("INSTALL_OTHER_PACKAGES")));
+        $cmd = $cmd . " $_" foreach (split(/,/, $additional_pkgs));
```
Comment (waynechen55) [inline tests/virt_autotest/transactional/prepare_transactional_server.pm:91]: Why this change ?

File: tests/virt_autotest/transactional/prepare_transactional_server.pm (hunk 13 above)
Comment (nanzhg) [reply tests/virt_autotest/transactional/prepare_transactional_server.pm:78]: I extract the common part as a function 'install_additional_pkgs()' for leverage in kubevirt tests with sle-micro. Actually, the variable 'INSTALL_OTHER_REPOS' is invoked 2 times, just want to state it at beginning to make it more explicit.
`if (get_var("INSTALL_OTHER_REPOS")) {`

File: tests/virt_autotest/transactional/prepare_transactional_server.pm (hunk 14 above)
Comment (nanzhg) [reply tests/virt_autotest/transactional/prepare_transactional_server.pm:91]: The same reason with the comment https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/15878#discussion_r1028045110

File: tests/virt_autotest/kubevirt_tests_server.pm (hunk 6 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm:209]: Yes, it works well in 15sp5 test. As we discussed offline, will add a judgment to check if manifests packages are in the default repo or not.

File: tests/virt_autotest/transactional/prepare_transactional_server.pm (hunk 13 above)
Comment (waynechen55) [reply tests/virt_autotest/transactional/prepare_transactional_server.pm:78]: It does not solve "calling twice" problem. In fact, any variable in a function will be called once when the function is called. So it looks to me this change is unnecessary.

Comment (nanzhg) [reply tests/virt_autotest/transactional/prepare_transactional_server.pm:78]: I'm ok if you'd like to keep the original code, no any impact for me.

File: tests/virt_autotest/kubevirt_tests_server.pm [hunk 15]
```diff
@@ -160,34 +181,49 @@ sub check_service_status {
...
     zypper_call("--gpg-auto-import-keys ref");
 
     my $virt_manifests = 'containerized-data-importer-manifests kubevirt-manifests kubevirt-virtctl';
+    my $search_manifests = $virt_manifests =~ s/\s+/\\\|/gr;
 
     if ($virt_manifests_repo) {
         zypper_call("in -f -r Virt-Manifests-Repo $virt_manifests");
     } elsif (script_run("rpmquery $virt_manifests")) {
-        zypper_call("in -f -r SLE-Module-Containers${os_version}-Updates $virt_manifests");
+        if (is_transactional || script_run("zypper se -r SLE-Module-Containers${os_version}-Updates $virt_manifests | grep -w '$search_manifests'")) {
```
Comment (alice-suse) [inline tests/virt_autotest/kubevirt_tests_server.pm:207]: Is the if condition given correct?  I guess it should be 
```if (is_transactional || ! script_run("zypper se -r SLE-Module-Containers${os_version}-Updates $virt_manifests | grep -w '$search_manifests'"))```

Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm:207]: Yes, it was verified. You can see it will return '0' if the repo included those packages.
Refer to http://10.67.129.66/tests/1735#step/kubevirt_tests_server/84

vh013:~ # zypper se -r SLE-Module-Containers15-SP5-Updates containerized-data-importer-manifests kubevirt-manifests kubevirt-virtctl |grep -w 'containerized-data-importer-manifests\|kubevirt-manifests\|kubevirt-virtctl'
i+ | containerized-data-importer-manifests | YAML manifests used to install CDI      | package
i+ | kubevirt-manifests                    | YAML manifests used to install kubevirt | package
i+ | kubevirt-virtctl                      | Client for managing kubevirt            | package
vh013:~ # echo $?
0

Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm:207]: Yes, it was verified. You can see it will return '0' if the repo included those packages.
Refer to http://10.67.129.66/tests/1735#step/kubevirt_tests_server/84

Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm:207]: Yes, it was verified. You can see it will return '0' if the repo included those packages.
Refer to http://10.67.129.66/tests/1735#step/kubevirt_tests_server/84

vh013:~ # zypper se -r SLE-Module-Containers15-SP5-Updates containerized-data-importer-manifests kubevirt-manifests kubevirt-virtctl |grep -w 'containerized-data-importer-manifests\|kubevirt-manifests\|kubevirt-virtctl'
i+ | containerized-data-importer-manifests | YAML manifests used to install CDI      | package
i+ | kubevirt-manifests                    | YAML manifests used to install kubevirt | package
i+ | kubevirt-virtctl                      | Client for managing kubevirt            | package
vh013:~ # echo $?
0

File: tests/virt_autotest/kubevirt_tests_server.pm [hunk 16]
```diff
@@ -104,6 +121,10 @@ sub rke2_server_setup {
         assert_script_run('curl -sfL https://get.rke2.io | sh -', timeout => 180);
     }
 
+    # Add kubectl command to $PATH environment varibable
+    my $rke2_bin_path = "export PATH=\$PATH:/opt/rke2/bin:/var/lib/rancher/rke2/bin";
+    assert_script_run("echo '$rke2_bin_path' > \$HOME/.bashrc; source \$HOME/.bashrc");
```
Comment (waynechen55) [inline tests/virt_autotest/kubevirt_tests_server.pm:126]: >> or > here in 
```
echo '$rke2_bin_path' > \$HOME/.bashrc
```
?

File: tests/virt_autotest/kubevirt_tests_server.pm (hunk 15 above)
Comment (alice-suse) [reply tests/virt_autotest/kubevirt_tests_server.pm:207]: Yep, sorry, my bad.

Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm:208]: That's ok. NP

File: tests/virt_autotest/kubevirt_tests_server.pm (hunk 16 above)
Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm]: To avoid the existing content lost in this file, it should use '>>' instead of '>'. Thanks for pointing out.

**Review by alice-suse — APPROVED** [review]
